File: study/ts/script/ui/PopupPlayerLevelUp.ts
const { ccclass, property } = cc._decorator;

import { AudioConst } from '../const/AudioConst';
import { ConfigNameConst } from '../const/ConfigNameConst';
import { DataConst } from '../const/DataConst';
import { SignalConst } from '../const/SignalConst';
import EffectGfxMoving from '../effect/EffectGfxMoving';
import EffectGfxNode from '../effect/EffectGfxNode';
import { Colors, G_AudioManager, G_ConfigLoader, G_ConfigManager, G_EffectGfxMgr, G_SignalManager, G_UserData } from '../init';
import { Lang } from '../lang/Lang';
import { UserDataHelper } from '../utils/data/UserDataHelper';
import { WayFuncDataHelper } from '../utils/data/WayFuncDataHelper';
import { handler } from '../utils/handler';
import { Path } from '../utils/Path';
import { TextHelper } from '../utils/TextHelper';
import UIActionHelper from '../utils/UIActionHelper';
import UIHelper from '../utils/UIHelper';
import { Util } from '../utils/Util';
import CommonContinueNode from './component/CommonContinueNode';
import CommonLevelUpAttr from './component/CommonLevelUpAttr';
import PopupBase from './PopupBase';
import PopupPlayerLevelUpCell from './PopupPlayerLevelUpCell';


@ccclass
export default class PopupPlayerLevelUp extends PopupBase {

    static preLoadRes = [
        'prefab/common/PopupPlayerLevelUpCell',
        'prefab/common/CommonLevelUpAttr'
    ];

    @property({
        type: cc.Node,
        visible: true
    })
    _panelRoot: cc.Node = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _panelBK: cc.Node = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _effectRootNode: cc.Node = null;

    @property({
        type: PopupPlayerLevelUpCell,
        visible: true
    })
    _wayGetCell1: PopupPlayerLevelUpCell = null;

    @property({
        type: PopupPlayerLevelUpCell,
        visible: true
    })
    _wayGetCell2: PopupPlayerLevelUpCell = null;

    @property({
        type: PopupPlayerLevelUpCell,
        visible: true
    })
    _wayGetCell3: PopupPlayerLevelUpCell = null;

    @property({
        type: CommonContinueNode,
        visible: true
    })
    _commonContinueNode: CommonContinueNode = null;
    @property({
        type: cc.ScrollView,
        visible: true
    })
    private _listView: cc.ScrollView = null;

    private _functionGuideList = [];
    private _fingerEffectNodeList = [];
    private _isAnimEnd = false;
    private _effectNode;
    private _callBack;

    onLoad() {
        super.onLoad();
        this._commonContinueNode.node.active = false;
        this._panelRoot.on(cc.Node.EventType.TOUCH_END, handler(this, this._onClickScreen));
    }

    setCloseCallback(closeCallback) {
        this._callBack = closeCallback;
    }

    _onClickScreen() {
        if (this._isAnimEnd) {
            // logWarn('PopupPlayerLevelUp:_onClickScreen');
            this.close();
        }
    }

    defaultUI() {
        var baseMgr = G_UserData.getBase();
        var roleCfgData = UserDataHelper.getUpgradeAttrData(baseMgr.getOldPlayerLevel(), baseMgr.getLevel());
        var newVit = baseMgr.getResValue(DataConst.RES_VIT);
        var newSpirit = baseMgr.getResValue(DataConst.RES_SPIRIT);
        // dump(roleCfgData);
        var oldVit = newVit - roleCfgData.lvup_power;
        if (oldVit < 0) {
            oldVit = 0;
        }
        if (oldVit == 0) {
            oldVit = newVit;
        }
        var oldSpirit = newSpirit - roleCfgData.lvup_energy;
        if (oldSpirit < 0) {
            oldSpirit = 0;
        }
        if (oldSpirit == 0) {
            oldSpirit = 0;
        }
        var attrList = [];
        attrList.push({
            desc: Lang.get('player_level_up'),
            oldValue: baseMgr.getOldPlayerLevel(),
            newValue: baseMgr.getLevel()
        });
        attrList.push({
            desc: Lang.get('player_level_vip'),
            oldValue: oldVit,
            newValue: newVit
        });
        attrList.push({
            desc: Lang.get('player_level_spirit'),
            oldValue: oldSpirit,
            newValue: newSpirit
        });
        this.updateUI(attrList);
        this._initLevelInfo();
        this._addEffectNode();
    }

    updateUI(params) {
        for (var i in params) {
            var value = params[i];
            var widget = this._createLineAttr();
            this._listView.content.addChild(widget.node);
            // this._listView.pushBackCustomItem(widget);
            this._setAttrLine(widget, value.desc, value.oldValue, value.newValue);
        }
        // this._addEffectNode();
    }

    _createLineAttr(): cc.Widget {
        var rootWidget = new cc.Node().addComponent(cc.Widget);
        // var listSize = this._listView.content.getContentSize();

        var advanceAttr: CommonLevelUpAttr = Util.getNode('prefab/common/CommonLevelUpAttr', CommonLevelUpAttr);
        var advanceRoot: cc.Node = Util.getSubNodeByName(advanceAttr.node, 'Panel_root');
        var advanceSize = advanceRoot.getContentSize();
        rootWidget.node.setContentSize(advanceSize);
        rootWidget.node.addChild(advanceAttr.node);
        UIHelper.setPosByPercent(advanceAttr.node, cc.v2(0.5, 0.5), null);
        rootWidget.node.active = false;
        return rootWidget;
    }

    _setAttrLine(attr: cc.Widget, desc, oldValue, newValue) {
        Util.updateLabelByName(attr.node, 'Text_desc', desc);
        oldValue = oldValue || 0;
        Util.updateLabelByName(attr.node, 'Text_value', oldValue);
        Util.updateLabelByName(attr.node, 'Text_up_value', newValue);
        var arrow = Util.getSubNodeByName(attr.node, 'Image_arrow');
        if (arrow) {
            UIActionHelper.playFloatXEffect(arrow);
        }
    }

    addAttr(desc, oldValue, newValue) {
        var widget = this._createLineAttr();
        this._listView.content.addChild(widget.node);
        // this._listView.pushBackCustomItem(widget);
        this._setAttrLine(widget, "attrId", oldValue, newValue);
    }

    _addEffectNode() {
        if (!this._effectNode) {
            G_AudioManager.playSoundWithId(AudioConst.SOUND_CREATE_ROLE_LEVEL_UP);
            this._effectNode = this._createEffectNode(this._effectRootNode);
        }
    }

    _onItemSelected(item, index) {
    }
    _onItemTouch(index, t) {
    }
    _onInit() {
    }
    onEnter() {
        G_SignalManager.dispatch(SignalConst.EVENT_TUTORIAL_START, "PopupPlayerLevelUp");
    }
    onClose() {
        if (this._callBack) {
            this._callBack(true);
        }
    }
    onExit() {
    }
    _initLevelInfo() {
        var infoList = [];
        var FunctionLevel = G_ConfigLoader.getConfig(ConfigNameConst.FUNCTION_LEVEL);
        var len = FunctionLevel.length();
        for (var i = 1; i <= len; i++) {
            var info = FunctionLevel.indexOf(i - 1);
            if (info && info.is_show_lvup != 0) {
                infoList.push(info);
            }
        }
        infoList.sort(function (a, b) {
            if (a.level == b.level) return 0;
            return a.level - b.level;
        });
        var maxShowNum = 2;
        var functionGuideList = this._getFunctionGuideList(infoList, maxShowNum);
        // console.log('>>>>>>> ', functionGuideList);
        var arr: string[] = [];
        for (var index in functionGuideList) {
            var value = functionGuideList[index];
            if (value.icon) arr.push(Path.getCommonIcon('main', value.icon));
        }
        this._functionGuideList = functionGuideList;
        if (arr.length > 0) {
            cc.resources.load(arr, cc.SpriteFrame, (err, resource) => {
                // console.log(err, resource);
                this.setFunctionGuideList();
            });
        }
        else {
            this.setFunctionGuideList();
        }
    }
    private setFunctionGuideList() {
        for (var index in this._functionGuideList) {
            var value = this._functionGuideList[index];
            this._updateWayInfo(parseInt(index) + 1, value);
        }
    }
    _updateWayInfo(index, wayInfo) {
        var wayItem: PopupPlayerLevelUpCell = this['_wayGetCell' + index];
        // console.log('>>>> ', index, wayItem);
        if (!wayItem) {
            // assert(false, 'PopupPlayerLevelUp:_updateWayInfo can\'t find wayItem by index : ' + index);
            return;
        }
        var playerLevel = G_UserData.getBase().getLevel();
        Util.updateImageView(wayItem._image_Icon, { texture: Path.getCommonIcon('main', wayInfo.icon), ignoreContentAdaptWithSize: true });
        var name = wayInfo.name;
        if (!G_ConfigManager.checkCanRecharge() && wayInfo.forbidIos == 1) {
             name = '' ;
        }
        Util.updateLabel(wayItem._text_way_name, { text: name });
        Util.updateLabel(wayItem._text_way_desc, { text: wayInfo.description });
        wayItem._button_OK.node.active = (playerLevel >= wayInfo.level && wayInfo.is_show_lvup == 1);
        var textDesc = Lang.get('way_type_no_open_desc', { level: wayInfo.level });
        Util.updateLabel(wayItem._text_way_unlock_desc, {
            visible: playerLevel < wayInfo.level,
            text: textDesc,
            color: Colors.getColor(6)
        });
        Util.updateImageView(wayItem._image_way_unlock, { visible: playerLevel >= wayInfo.level && wayInfo.is_show_lvup == 2 });
        var canSkip = playerLevel >= wayInfo.level && wayInfo.is_show_lvup == 1;
        if (wayItem._panel_Touch && canSkip) {
            wayItem._panel_Touch.on(cc.Node.EventType.TOUCH_END, () => {
                if (!this._commonContinueNode.node.active) return;
                if (canSkip) {
                    WayFuncDataHelper.gotoModuleByFuncId(wayInfo.function_id);
                }
                // this.close();
            });
        } else {
            wayItem._panel_Touch.active = false;
        }
        // console.log('>> ', playerLevel < wayInfo.level, playerLevel >= wayInfo.level && wayInfo.is_show_lvup == 2);
        // wayItem.node.active = false;
        if (wayInfo.is_show_lvup == 1 && wayInfo.is_guide == 1) {
            var effectNode = new EffectGfxNode();
            effectNode.setEffectName('effect_finger');
            wayItem._button_OK.node.addChild(effectNode.node);
            effectNode.play();
            UIHelper.setPosByPercent(effectNode.node, cc.v2(0.5, 0.5), null);
            this._fingerEffectNodeList.push(effectNode);
            // effectNode.node.active = false;
        }
    }

    _getFunctionGuideList(infoList, maxShowNum) {
        var lastIndex = null;
        for (var i = 0; i < infoList.length; i++) {
            var info = infoList[i];
            if (info.level >= G_UserData.getBase().getLevel()) {
                lastIndex = i;
                break;
            }
        }
        lastIndex = lastIndex == null ? (infoList.length - 1) : lastIndex;
        var start = lastIndex;
        var dstInfos = [];
        do {
            dstInfos.push(infoList[start]);
            // dstInfos[dstInfos.length] = infoList[start];
            start = start + 1;
        } while (dstInfos.length < maxShowNum && infoList[start] != null);
        if (dstInfos.length < maxShowNum) {
            start = lastIndex - 1;
            do {
                dstInfos.push(infoList[start]);
                // dstInfos[dstInfos.length] = infoList[start];
                start = start - 1;
            } while (dstInfos.length < maxShowNum && infoList[start] != null);
        }
        // assert(dstInfos.length >= maxShowNum, 'There is not enough info with open level: ' + tostring(G_UserData.getBase().getLevel()));
        dstInfos.sort((a, b) => {
            if (a.level == b.level) return 0;
            return a.level - b.level;
        });
        return dstInfos;
    }
    isAnimEnd() {
        return this._isAnimEnd;
    }
    _createEffectNode(effectRootNode) {
        let effectFnc = function (event) {
            // console.log('event: ', event);
            if (TextHelper.stringStartsWith(event, 'play_shuzhi')) {
                var index = TextHelper.stringGetSuffixIndex(event, 'play_shuzhi');
                var item = this._listView.content.children[index - 1];
                if (item) {
                    item.active = true;
                    G_EffectGfxMgr.applySingleGfx(item, 'smoving_herolevelup_shuzhi', null, null, null);
                }
            } else if (TextHelper.stringStartsWith(event, 'play_zhaomu')) {
                var index = TextHelper.stringGetSuffixIndex(event, 'play_zhaomu');
                if (index > this._functionGuideList.length) {
                    return;
                }
                var wayGetCell = this['_wayGetCell' + index];
                if (wayGetCell) {
                    wayGetCell.node.active = true;
                    G_EffectGfxMgr.applySingleGfx(wayGetCell.node, 'smoving_herolevelup_zhaomu', null, null, null);
                }
            } else if (event == 'play_dianji') {
                this._commonContinueNode.node.active = (true);
            } else if (event == 'finish') {
                for (var k in this._fingerEffectNodeList) {
                    var v = this._fingerEffectNodeList[k];
                    v.node.active = true;
                }
                this._isAnimEnd = true;
            }
        }
        var moving: EffectGfxMoving = G_EffectGfxMgr.createPlayMovingGfx(effectRootNode, 'moving_herolevelup', null, effectFnc.bind(this), false);
        return moving.node;
    }
}

File: study/lua/src/app/ui/PopupPlayerLevelUp.lua
--弹出界面
--主公等级提升
--
local PopupBase = require("app.ui.PopupBase")
local PopupPlayerLevelUp = class("PopupPlayerLevelUp", PopupBase)
local Path = require("app.utils.Path")
local UIHelper  = require("yoka.utils.UIHelper")
local CSHelper  = require("yoka.utils.CSHelper")
local TypeConvertHelper = require("app.utils.TypeConvertHelper")
local WayFuncDataHelper = require("app.utils.data.WayFuncDataHelper")
local DataConst = require("app.const.DataConst")
local UserDataHelper = require("app.utils.UserDataHelper")
local AudioConst = require("app.const.AudioConst")
local LINE_ITEM_COUNT = 5 --一行5个ICON
local LINE_HEIGHT = 130
local LINE_ITEM_BLACK = 3 -- 物品横排间隔


function PopupPlayerLevelUp:ctor(closeCallback)
	self._listView  = nil
	--根据玩家等级开放功能
	self._wayGetCell1 = nil
	self._wayGetCell2 = nil
	self._wayGetCell3 = nil
	self._effectRootNode = nil
	self._callBack = closeCallback
	self._functionGuideList = {}
	self._effectNode = nil
	self._fingerEffectNodeList = {}
	self._panelRoot = nil
	self._isAnimEnd = false
	local resource = {
		file = Path.getCSB("PopupPlayerLevelUp", "common"),
		binding = {
			_panelRoot = {
				events = {{event = "touch", method = "_onClickScreen"}}
			}
		},
	}

	self:setName("PopupPlayerLevelUp")
	PopupPlayerLevelUp.super.ctor(self, resource, false, false)
end


function PopupPlayerLevelUp:_onClickScreen()
	if self._isAnimEnd then
		logWarn("PopupPlayerLevelUp:_onClickScreen")
		self:close()
	end
end

function PopupPlayerLevelUp:defaultUI()
	local baseMgr = G_UserData:getBase()
	local roleCfgData = UserDataHelper.getUpgradeAttrData(baseMgr:getOldPlayerLevel(),baseMgr:getLevel())
	local newVit =  baseMgr:getResValue(DataConst.RES_VIT)
	local newSpirit = baseMgr:getResValue(DataConst.RES_SPIRIT)
	dump(roleCfgData)
	local oldVit = newVit - roleCfgData.lvup_power
	if oldVit < 0 then
		oldVit = 0
	end
	if oldVit == 0 then
		oldVit = newVit
	end
	
	
	local oldSpirit  = newSpirit - roleCfgData.lvup_energy
	if oldSpirit < 0 then
		oldSpirit = 0
	end
	if oldSpirit == 0 then
		oldSpirit = 0
	end
    local attrList = {
        [1] = {desc = Lang.get("player_level_up"), oldValue = baseMgr:getOldPlayerLevel(), newValue = baseMgr:getLevel()},
        [2] = {desc = Lang.get("player_level_vip"), oldValue = oldVit, newValue = newVit},
        [3] = {desc = Lang.get("player_level_spirit"), oldValue = oldSpirit, newValue = newSpirit},
    }
	self:updateUI(attrList)
	self:_initLevelInfo()
	self:_addEffectNode()
end


function PopupPlayerLevelUp:updateUI(params)
	for i, value in ipairs(params) do
		if checktable(value) then
			local widget = self:_createLineAttr()
			self._listView:pushBackCustomItem(widget)
			self:_setAttrLine(widget, value.desc, value.oldValue, value.newValue)
		end
	end
--	self._listView:adaptWithContainerSize()
	self:_addEffectNode()
end

function PopupPlayerLevelUp:_createLineAttr()
	local rootWidget = ccui.Widget:create()
	local listSize = self._listView:getContentSize()
	
	local advanceAttr = CSHelper.loadResourceNode(Path.getCSB("CommonLevelUpAttr", "common"))
	local advanceRoot = ccui.Helper:seekNodeByName(advanceAttr, "Panel_root")
	local advanceSize = advanceRoot:getContentSize()
	rootWidget:setContentSize(advanceSize)
	rootWidget:addChild(advanceAttr)
	UIHelper.setPosByPercent(advanceAttr,cc.p(0.5,0.5))

	rootWidget:setVisible(false)
	return rootWidget
end

function PopupPlayerLevelUp:_setAttrLine(attr, desc, oldValue, newValue)
	attr:updateLabel("Text_desc", desc)

	oldValue = oldValue or 0
	attr:updateLabel("Text_value", oldValue)

	attr:updateLabel("Text_up_value", newValue)

	local arrow = attr:getSubNodeByName("Image_arrow")
	if arrow then
		local UIActionHelper = require("app.utils.UIActionHelper")
		UIActionHelper.playFloatXEffect(arrow)
	end
end

function PopupPlayerLevelUp:addAttr( desc, oldValue, newValue)
	local widget = self:_createLineAttr()

	self._listView:pushBackCustomItem(widget)

	self:_setAttrLine(widget, attrId, oldValue, newValue)

	--self._listView:adaptWithContainerSize()
end



--

function PopupPlayerLevelUp:onCreate()
	-- button
	self._listView = ccui.Helper:seekNodeByName(self, "ListView")
	self._listView:setVisible(true)
	self._commonContinueNode:setVisible(false)
	
end

function PopupPlayerLevelUp:_addEffectNode()
	if not self._effectNode then
		G_AudioManager:playSoundWithId(AudioConst.SOUND_CREATE_ROLE_LEVEL_UP)
		self._effectNode = self:_createEffectNode(self._effectRootNode )
	end
end

function PopupPlayerLevelUp:_onItemSelected(item, index)
end

function PopupPlayerLevelUp:_onItemTouch(index, t)
end

function PopupPlayerLevelUp:_onInit()

end


function PopupPlayerLevelUp:onEnter()
   	G_SignalManager:dispatch(SignalConst.EVENT_TUTORIAL_START, self.__cname)
	--G_SignalManager:dispatch(SignalConst.EVENT_TUTORIAL_TOUCH_AUTH_BEGIN)
end

function PopupPlayerLevelUp:onClose( ... )
	-- body
	--G_SignalManager:dispatch(SignalConst.EVENT_TUTORIAL_TOUCH_AUTH_END)
	--G_SignalManager:dispatch(SignalConst.EVENT_TUTORIAL_STEP,self.__cname)
	if self._callBack then
		self._callBack(true)
	end
end

function PopupPlayerLevelUp:onExit()

end

--开区功能
function PopupPlayerLevelUp:_initLevelInfo()
	local infoList = {}
	local FunctionLevel = require("app.config.function_level")
	for i=1, FunctionLevel.length() do 
		local info = FunctionLevel.indexOf(i)
		if info.is_show_lvup ~= 0 then
			table.insert(infoList, info)
		end
	end
	-- 按照等级从小到大排序
	table.sort(infoList, function(a, b)
		return a.level < b.level
	end)
	
	local maxShowNum = 2
	local functionGuideList = self:_getFunctionGuideList(infoList,maxShowNum)
	for index, value  in ipairs(functionGuideList) do
		self:_updateWayInfo(index, value)
	end
	self._functionGuideList  = functionGuideList
end

function PopupPlayerLevelUp:_updateWayInfo(index, wayInfo)
	local wayItem = self["_wayGetCell"..index]
	if wayItem == nil then
		assert(false, "PopupPlayerLevelUp:_updateWayInfo can't find wayItem by index : "..index)
		return
	end
	local playerLevel =  G_UserData:getBase():getLevel() 
	wayItem:updateImageView("Image_Icon", {texture = Path.getCommonIcon("main",wayInfo.icon)})
	wayItem:updateLabel("Text_way_name", {text = wayInfo.name})--color = Colors.COLOR_SCENE_DESC_NORMAL
	wayItem:updateLabel("Text_way_desc", {text = wayInfo.description})--color = Colors.COLOR_SCENE_TIP

	if playerLevel  >= wayInfo.level and wayInfo.is_show_lvup == 1  then
		wayItem:updateButton("Button_OK", {visible = true })
	else
		wayItem:updateButton("Button_OK", {visible = false })
	end

	local textDesc = Lang.get("way_type_no_open_desc", {level = wayInfo.level} )

	wayItem:updateLabel("Text_way_unlock_desc", 
	{
		visible = playerLevel < wayInfo.level, 
		text = textDesc, 
		color = Colors.getColor(6), 
	})

    local textDesc = Lang.get("way_type_open")
	wayItem:updateLabel("Text_way_unlock_desc", 
	{
		visible = playerLevel >= wayInfo.level and wayInfo.is_show_lvup == 2, 
		text = textDesc, 
		color = Colors.getColor(2), 
    })
    
	wayItem:updateImageView("Image_way_unlock", 
	{
		visible = false,--playerLevel >= wayInfo.level and wayInfo.is_show_lvup == 2, 
	})


	local panelTouch = wayItem:getSubNodeByName("Panel_Touch")
	local canSkip = playerLevel >= wayInfo.level and wayInfo.is_show_lvup == 1
	if panelTouch and canSkip then
		local function onGoHandler()
			logWarn("click panelTouch"..wayInfo.level.."_________"..wayInfo.is_show_lvup)
			if canSkip then
				-- 获取入口，然后切换，未来考虑把入口直接进入进ModuleDirector里
				WayFuncDataHelper.gotoModuleByFuncId(wayInfo.function_id)
				self:close()
			else
				self:close()	
			end
		end
		
		panelTouch:setTouchEnabled(true)
		panelTouch:addClickEventListenerEx(onGoHandler)
	else
		panelTouch:setTouchEnabled(false)	
	end
	wayItem:setVisible(false)

	if wayInfo.is_show_lvup == 1 and  wayInfo.is_guide == 1 then
		local buttonOk = ccui.Helper:seekNodeByName(wayItem, "Button_OK")
		local EffectGfxNode = require("app.effect.EffectGfxNode")
		local effectNode = EffectGfxNode.new("effect_finger")
		buttonOk:addChild(effectNode)
		effectNode:play()
		UIHelper.setPosByPercent(effectNode, cc.p(0.5,0.5))
		table.insert( self._fingerEffectNodeList, effectNode)
		 effectNode:setVisible(false)--整个动画播放完才显示
	end
end

function PopupPlayerLevelUp:_getFunctionGuideList(infoList,maxShowNum)

	-- 先找到当前大于等于当前等级条目数据的索引
	local lastIndex = nil
	for i=1, #infoList do
		local info = infoList[i]
		if info.level >= G_UserData:getBase():getLevel() then
			lastIndex = i
			break
		end
	end
	-- 有可能没有匹配的，这时候说明是已经满级了
	lastIndex = lastIndex or #infoList

	-- 往下找三个
	local start = lastIndex
	local dstInfos = {}
	repeat
		dstInfos[#dstInfos + 1] = infoList[start]
		start = start + 1
	until #dstInfos >= maxShowNum or not infoList[start]

	-- 如果还不满足，则往上找三个
	if #dstInfos < maxShowNum then
		local start = lastIndex - 1	-- 这里-1是因为lastIndex已经在上面被加过了
		repeat
			dstInfos[#dstInfos + 1] = infoList[start]
			start = start - 1
		until #dstInfos >= maxShowNum or not infoList[start]
	end

	-- 理论上还有可能不满足的，这时候就说明配置有问题了
	assert(#dstInfos >= maxShowNum, "There is not enough info with open level: "..tostring(G_UserData:getBase():getLevel()))

	-- 按照等级从小到大排序
	table.sort(dstInfos, function(a, b)
		return a.level < b.level
	end)

	return dstInfos
end

function PopupPlayerLevelUp:isAnimEnd()
	return self._isAnimEnd
end
function PopupPlayerLevelUp:_createEffectNode(effectRootNode)
	local TextHelper = require("app.utils.TextHelper")
    local EffectGfxNode = require("app.effect.EffectGfxNode")
    local function effectFunction(effect)
        if TextHelper.stringStartsWith(effect,"effect_") then
            local subEffect = EffectGfxNode.new(effect)
            subEffect:play()
            return subEffect
		else
			return display.newNode()  
		end
    end
    local function eventFunction(event,frameIndex,movingNode)
        if TextHelper.stringStartsWith(event,"play_shuzhi") then
			local index = TextHelper.stringGetSuffixIndex(event,"play_shuzhi")
			local items = self._listView:getItems()
			if items[index] then
				items[index]:setVisible(true)
				G_EffectGfxMgr:applySingleGfx(items[index], "smoving_herolevelup_shuzhi", nil, nil, nil)
			end
		elseif TextHelper.stringStartsWith(event,"play_zhaomu") then	
            local index = TextHelper.stringGetSuffixIndex(event,"play_zhaomu")
            if index == nil then
                return
            end
			if #self._functionGuideList < index then
				return 
			end
			local wayGetCell = self["_wayGetCell"..index]
			if wayGetCell then
				wayGetCell:setVisible(true)
				G_EffectGfxMgr:applySingleGfx(wayGetCell, "smoving_herolevelup_zhaomu", nil, nil, nil)
			end
		elseif event == "play_dianji"  then	 	
			self._commonContinueNode:setVisible(true)
		elseif event == "finish" then	 	 	
			for k,v in ipairs(self._fingerEffectNodeList) do
				 v:setVisible(true)
			end
			self._isAnimEnd  = true
        end
    end
   local node =  G_EffectGfxMgr:createPlayMovingGfx( effectRootNode, "moving_herolevelup", effectFunction, eventFunction , false )
   return node
end


return PopupPlayerLevelUp